### DISCUSSION TITLE
Vaultwarden API 500: Internal Server Error

### DISCUSSION THREAD

**Discussion by modarken** [dani-garcia/vaultwarden#2993] [Ideas]
### Subject of the issue
Using Postman to try and connect to vaultwarden API I get a 500: Internal Server Error

### Deployment environment
Docker

* vaultwarden version: 1.26

* Install method:
* docker pull
* 
* Clients used: 
* chrome

* Other relevant details:

### Steps to reproduce
POST /public/identity/connect/token HTTP/1.1
Host: bitwarden.test.local
Content-Type: application/x-www-form-urlencoded
Content-Length: 136

grant_type=client_credentials&scope=api&client_id=user.xxxxx&client_secret=6Yxxxxx

### Expected behavior
json return of token

### Actual behaviour
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>500 Internal Server Error</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>500: Internal Server Error</h1>
        <p>The server encountered an internal error while processing this request.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>

**Comment by BlackDex** [comment]
Vaultwarden doesn't support the Bitwarden Public API, so that will never work.
At this moment Vaultwarden only supports the client API, which is why your request fails.

If you really want to login this way, make sure to use the client API, which also provides a `device_identifier` and `device_name`
The `device_identifier` is a v4 UUID, which stays unique per device/client and keeps the same per user, even if it is the exact same device/application.

**Comment by uedvt359** [comment]
this is the endpoint for retrieving bearer tokens for the *vault management api* (which vaultwarden supports), not the *public api*. 

the problem is that one must pass additional parameters [`device_identifier=aaa&device_name=bbb&device_type=ccc`](https://github.com/dani-garcia/vaultwarden/commit/1ca0d6e2451ae6f68b0325e7ae9f1233185bbb68). the values don't seem to matter, and there seems to be no documentation about what they should contain.

**Reply by BlackDex** [reply]
That is not entirely true.

The `device_type` is defined here https://github.com/bitwarden/server/blob/c08e2a747367749eb897082ddd70f867bafe307f/src/Core/Enums/DeviceType.cs as stated here https://github.com/dani-garcia/vaultwarden/blob/445680fb84d238aa432fdec271235933043410e4/src/db/models/device.rs#LL18C28-L18C104

The `device_name` is a string which represents the name of the client used, if you would check the database you will see different names like, `firefox`, `chrome`, `linux`. But you can put in there what you want to better identify the client used.

And, the same goes for `device_identifier`, that is just a [UUIDv4](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)), which is all stated in code.

**Comment by uedvt359** [comment]
This discussion was closed, but should this return e.g. a 400 Bad Request instead of a 500 Internal Server Error?
